vars/stepResult.groovy

```diff
@@ -22,112 +22,108 @@ def call(Map config= [:]) {
-
-        if (!config['ignore_failure']) {
-            currentBuild.result = config.get('result')
-        }
-
-        if (config['result'] == "ABORTED" ||
-            config['result'] == "UNSTABLE" ||
-            config['result'] == "FAILURE") {
-             def comment_url = env.BUILD_URL + "display/redirect"
-
-            if (env.CHANGE_ID) {
-                if (log_url) {
-                    comment_url = log_url
-                }
-
-                pullRequest.comment("Test stage ${config.name}" +
-                                    " completed with status " +
-                                    "${config.result}" +
-                                    ".  " + comment_url)
-            }
-        }
-
-        def result = config['result']
-        switch(config['result']) {
-            case "UNSTABLE":
-                result = "FAILURE"
-                break
-            case "FAILURE":
-                result = "ERROR"
-                break
-        }
-        Map param =[:]
-        param['description'] = config['name']
-        if (config['context'].contains('/')) {
-            // Allow migration to common context value for scmNotify and
-            // stepResult
-           param['context'] = config['context']
-        } else {
-           param['context'] = config['context'] + "/" + config['name']
-        }
-        if (log_url) {
-           param['targetURL'] = log_url
-        }
-        param['status'] = result
-
-        scmNotify param
+  Map param =[:]
+  param['description'] = config['name']
+  if (config['context'].contains('/')) {
+    // Allow migration to common context value for scmNotify and
+    // stepResult
+    param['context'] = config['context']
+  } else {
+    param['context'] = config['context'] + "/" + config['name']
+  }
+
+  def log_url = null
+
+  if (env.DAOS_JENKINS_NOTIFY_STATUS == null) {
+    println "Jenkins not configured to notify github of builds."
+    return
+  }
+
+  if (config['junit_files'] && config['result'] != 'FAILURE') {
+    log_url = env.JOB_URL - ~/job\/[^\/]*\/$/ +
+              "/view/change-requests/job/${env.BRANCH_NAME}/" +
+              "${env.BUILD_ID}/testReport/(root)/"
+  } else {
+
+    def h = new com.intel.doGetHttpRequest()
+    resp = h.doGetHttpRequest(env.JOB_URL - ~/\/job\/[^\/]*\/$/ +
+           "/view/change-requests/job/" +
+    env.BRANCH_NAME.replaceAll('/', '%252F') +
+      "/${env.BUILD_ID}/wfapi/describe");
+
+    // def job = parseResponse(resp)
+    def job = readJSON text: resp
+    assert job instanceof Map
+
+    def stage
+    for (s in job['stages']) {
+      if (s['name'] == env.STAGE_NAME) {
+        stage = s
+        break
+      }
+    }
+    resp = h.doGetHttpRequest("${env.JENKINS_URL}" +
+           stage['_links']['self']['href'])
+
+    stage = readJSON text: resp
+    assert stage instanceof Map
+
+    def stageFlowNode = null
+    for (s in stage['stageFlowNodes']) {
+      if (s['name'].contains(env.STAGE_NAME)) {
```

Comment:
We need to pick our stage names carefully to prevent such cases.

Jenkins does not always use the exact name stage name in the names returned in stage['stageFlowNodes'], and I was getting failures during testing until I temporarily added diagnostics to print out all the names being returned.

Reply:
> We need to pick our stage names carefully to prevent such cases.

But we are trying to promote the existence of the `Jenkinsfile` as accessible to lesser-Jenkins-experienced developers (and it is working!) in a effort to allow them to help themselves instead of relying on us to do everything.  Achieving that means elminating this kind of _tribal knowledge_.  It should just be as easy as them picking whatever name they want without knowing they cannot have super/subset collisions, etc.

> Jenkins does not always use the exact name stage name in the names returned in stage['stageFlowNodes'],

But it must [have] or the previous use of `==` instead of `contains()` would have run into the same problems.  Are you sure this was not just a case of needing to `trim()` values of trailing whitespace or somesuch?

> and I was getting failures during testing until I temporarily added diagnostics to print out all the names being returned.

So what kind of "non-exact stage name" mappings were being seen with that?  What was different about our names vs. what was in `stage['stageFlowNodes']`.  At minimum, it has to be determinate such that we can construct the same name.

Reply:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-2743/24/pipeline

The env.STAGE_NAME is 'Build RPM on CentOS 7'

In the above case, the name returned in the list of 'stageFlowNodes' that is the closest match appears to be the name that used as a context variable by a previous call to githubNotify (wrapped by scmNotify)

So far I have not found documentation on how Jenkins names the flow-node.
And I did not notice previously that it chose the name that was previously passed as a context to githubNotify.

The naming may have changed with removing the enclosing node() for this step.

Reply:
As an example, given https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-2743/78/wfapi/, if we look at the stage:

```
{
    "_links": {
        "self": {
            "href": "/job/daos-stack/job/daos/view/change-requests/job/PR-2743/78/execution/node/112/wfapi/describe"
        }
    },
    "id": "112",
    "name": "Build RPM on CentOS 7",
    "execNode": "",
    "status": "SUCCESS",
    "startTimeMillis": 1592423750463,
    "durationMillis": 882810,
    "pauseDurationMillis": 0
},
```
and follow it's `_links`->`self` -> `href` to https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-2743/78/execution/node/112/wfapi/describe and then look through the steps there we can find:

```
{
    "_links": {
        "self": {
            "href": "/job/daos-stack/job/daos/view/change-requests/job/PR-2743/78/execution/node/192/wfapi/describe"
        },
        "log": {
            "href": "/job/daos-stack/job/daos/view/change-requests/job/PR-2743/78/execution/node/192/wfapi/log"
        },
        "console": {
            "href": "/job/daos-stack/job/daos/view/change-requests/job/PR-2743/78/execution/node/192/log"
        }
    },
    "id": "192",
    "name": "build/Build RPM on CentOS 7",
    "execNode": "",
    "status": "SUCCESS",
    "parameterDescription": " TARGET=centos7 ci/rpm/build.sh",
    "startTimeMillis": 1592423758159,
    "durationMillis": 865005,
    "pauseDurationMillis": 0,
    "parentNodes": [
        "191"
    ]
},
```

So the name we are looking for is actually _build/Build RPM on CentOS 7_ so it's our stage name prefixed with _build/_. So the question is where is the build/_ prefix coming from?  And yes, it looks like it's our GitHub status context.

Is it not always our _build/_ context such that we can do:

```suggestion
      if (s['name'] == 'build/' + env.STAGE_NAME)) {
```

or at worst do:

```suggestion
      if (s['name'].endsWith('/' + env.STAGE_NAME)) {
```
To know that we are matching on the whole stage name and not just a substring of it?

The former would be preferable, unless we know that the prefix won't always be the _build_ context, but I cannot imagine why it would not be.  Or maybe the function should take the context prefix as an argument to avoid hard-coding values that could change.  We'd then need to figure out how to get the context in the caller, or perhaps the caller is the right place to hard-code the context value since it's the caller that is defining it.

Reply:
What Jenkins documentation I can find on it claimed it has two features:
1. It is human readable.
2. It is unique for that stage.

he stage flow name for shell scripts comes from the label that is passed to "sh" steps.

I looked at the names dumped and they are almost a 1 for 1 correspondence the titles of the steps listed for that stage in the Web UI.

I just happen to use the context string for the shell script label for the RPM build instead of just using the just the Stage name as a simplification.

And the dump of the stage flow names does not show them to be unique.

So it looks like another parameter to pass to stepResult with a default of env.STAGE_NAME for backwards compatibility.